k2/csrc/array.h

```diff
@@ -189,6 +189,46 @@ struct Array2 {
  */
 };
 
+// Specialized Array2 for std::pair*, we represents something like
+// vector<vector<std::pair<U,V>> as an Array2 object with two data members
+// `U* data1` and `V* data2`. Given any `indexes[i]`,
+// `data1[indexes[i]]` and `data2[indexes[i]]` can be viewed as a pair
+// in the original vector of vector<std::pair<U,V>>.
+template <typename U, typename V, typename I>
+struct Array2<std::pair<U, V> *, I> {
```

Comment:
Note I specialize `Array2` for `std::pair` with two data members (`data1` and `data2`), so that there's no need to cast type when creating pybindings for `std::pair` (e.g. `LogSumTracebackState::DerivType`)

Reply:
Sorry, but I don't think I like this design.  The problem is that this Array2 looks like a regular Array2 but doesn't behave like one; it doesn't have the same interface or the same members.  It's templated on the pair, but how do you access the pair type?  I don't see any data-member that points to it.

I think it would be better to have the underlying data be a float array (say), but reinterpret the memory as int where needed.

Reply:
OK, let's say we have `arc_derivs = tensor.FloatTensor([num_arcs_derivs, 2])` on Python side, and `Array2<std::pair<int32_t, float>*` on C++ side. Suppose we set `arc_id = 1` on C++ side, then we'll get value like `1.4013e-45`  on Python side as a float. As most of algorithms accept `int*` or `Array1<int32*>` as `arc_map`, we need to reinterpret the python's float memory as int on C++ side  every time when we interact with code in C++, is that right? However, as we will wrap those algorithms to Python, so on Python side, those algorithms will accpet `IntTensor`, if we want to pass the first column of `arc_derivs` (i.e. `arc_derivs[: 0]` to get arc-ids) above to those Python algorithms, we need to convert `tensor.FloatTensor` as `tensor.IntTensor`,right?   I just worry about the accessibility for users and the scalability for other frameworks.

RE  the interface, if we want it behave like `Array2`, i.e. access `data1` and `data2` as pair type, maybe we can make `data1` and `data2` as private, and define some getter and setter like `void set_data(std::pair<int, float>, int32_t index)`? Well, still not like `Array2` but make it easy to interact with other algorithms in Python code.

Actually, I propose that we create a new struct of Array2 (maybe a new name instead of `Array2`) as the specialized version of `std::pair`, i.e. we have `IndexT indexes` in this version `Array2<some_tuple_type/some_struct_type, I>`, but will map all fields in type `some_tuple_type/some_struct_type` to mutiple `data` (`data1, data2, data3...`) in this `Array2`. So there's no need to reinterpret or cast in any framework.

Reply:
In general we're going to have these kinds of problems, e.g. in the cuda stuff I am putting a float field in Arc so it has mixed int and float fields.  But I think the best way to handle it is to have it of a fixed arbitrary type on the Python side (e.g. float here) and cast as needed.  These will largely be used in side k2 algorithms as temporaries, user code won't, generally speaking, need to access these kinds of things.  So I don't think it really matters.  Anyway we can wrap them in some kind of class on the python side that knows how to deal with them, if needed.  So just do it the way I said, for now.  I think you are over-thinking this.

Reply:
ok